engine/collection/pusher/engine.go

```diff
@@ -53,55 +83,86 @@ func New(log zerolog.Logger, net network.EngineRegistry, state protocol.State, e
 	}
 	e.conduit = conduit
 
+	e.cm = component.NewComponentManagerBuilder().
+		AddWorker(e.inboundMessageWorker).
+		Build()
+	e.Component = e.cm
+
 	return e, nil
 }
 
-// Ready returns a ready channel that is closed once the engine has fully
-// started.
-func (e *Engine) Ready() <-chan struct{} {
-	return e.unit.Ready()
+func (e *Engine) inboundMessageWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	done := ctx.Done()
+	wake := e.notifier.Channel()
+	for {
+		select {
+		case <-done:
+			return
+		case <-wake:
+			e.processInboundMessages(ctx)
+		}
+	}
 }
 
-// Done returns a done channel that is closed once the engine has fully stopped.
-func (e *Engine) Done() <-chan struct{} {
-	return e.unit.Done()
+func (e *Engine) processInboundMessages(ctx context.Context) {
+	for {
+		nextMessage, ok := e.inbound.Pop()
+		if !ok {
+			return
+		}
+
+		asEngineWrapper := nextMessage.(*engine.Message)
+		asSCGMsg := asEngineWrapper.Payload.(*messages.SubmitCollectionGuarantee)
+		originID := asEngineWrapper.OriginID
+
+		_ = e.process(originID, asSCGMsg)
+
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+	}
 }
 
 // SubmitLocal submits an event originating on the local node.
 func (e *Engine) SubmitLocal(event interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(e.me.NodeID(), event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
+	err := e.messageHandler.Process(e.me.NodeID(), event)
+	if err != nil {
+		engine.LogError(e.log, err)
+	}
 }
 
 // Submit submits the given event from the node with the given origin ID
 // for processing in a non-blocking manner. It returns instantly and logs
 // a potential processing error internally when done.
 func (e *Engine) Submit(channel channels.Channel, originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(originID, event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
+	err := e.messageHandler.Process(originID, event)
+	if err != nil {
+		engine.LogError(e.log, err)
+	}
 }
 
 // ProcessLocal processes an event originating on the local node.
```

Comment:
Instead of improving the implementation of the deprecated `Engine` interface, I think we should change the implementation to an application-specific interface (see [this](https://github.com/onflow/flow-go/pull/6747#discussion_r1857537372) and [this](https://github.com/onflow/flow-go/pull/6747#discussion_r1857672727) comment)

Reply:
`MessageHandler` no longer being used (https://github.com/onflow/flow-go/pull/6747/commits/f66ba69f012d29842c3b3a9f631b7a43d83088a1). Application-specific interface will be created as part of the following PR.